Added sampling methods of Discrete Random Variables

Smit-create opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
Added sampling methods of Discrete Random Variables
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
#17257

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * Added sampling methods for discrete random variables
<!-- END RELEASE NOTES -->

Smit-create committed:
sampling of drv

Smit-create committed:
fixing scipy not found

Smit-create committed:
clean ups

Smit-create commented:
@czgdp1807  Please review this

Smit-create commented:
@czgdp1807  Is this ready to merge or it may require some changes?

czgdp1807 commented on sympy/stats/drv.py:
```diff
@@ -45,8 +46,10 @@ def __new__(cls, *args):
         pass
 
-    def sample(self):
+    def sample(self,size=()):
```
Please provide some examples of using this in your PR in a reply to this comment.

czgdp1807 commented:
What about providing support for tensorflow?

czgdp1807 commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -261,3 +262,48 @@ def test_product_spaces():
+        skip('PyMC3 is not installed. Abort tests for _sample_pymc3.')
+    else:
+        for X in distribs_pymc3:
+            sam = X.pspace.distribution._sample_pymc3(size)
```
Does `sam = sample(X, size)` works?

Smit-create replied:
Yes, it works

czgdp1807 replied:
Please update the tests to use something like, `sam = sample(X, size)`.

Smit-create replied:
I have used `samps = sample(X, size) `in case of scipy tests and not in others because it covers almost all distribution but on the other hand numpy and pymc3 does not, so for scipy tests I have used `samps = sample(X, size)` and for other two  `samps = X.pspace.distribution._sample_numpy(size)` and `samps = X.pspace.distribution._sample_pymc3(size)`

Smit-create replied:
> Please update the tests to use something like, `sam = sample(X, size)`

I also added tests for using size wherever I changed the parameters

Smit-create committed:
Merge branch 'master' into sample_drv

Smit-create committed:
Added sample with size argument in stats

Smit-create committed:
Documentation of sample

Smit-create committed:
use sample instead of sample_scipy

Smit-create committed:
remove unsed import

Smit-create commented:
> What about providing support for tensorflow?

Tensorflow provides backend to PYMC3, so should we check the tensorflow support using,
`if import_module('tensorflow')` while using pymc3 model?

Smit-create commented:
@czgdp1807 It is passing the tests locally. But on travis it fails one test, please can you look into it?

Smit-create committed:
removed sample poisson without scipy

Smit-create commented:
Changes I made:
* Added option of `size` in `sample` method.
* Removed `sample` method in `PoissonDistribution` along with the failing tests without `scipy`.
* `sample` now calls `_sample_scipy` for 5 distributions directly.

Smit-create commented:
@czgdp1807  Is it good now?

Smit-create commented:
@czgdp1807 Please review.

czgdp1807 commented:
Please merge master as the coverage reports are using some old diffs.

Smit-create commented:
I am merging it.

Smit-create committed:
Merge branch 'master' into sample_drv

Smit-create committed:
added more tests

Smit-create commented:
@czgdp1807  Is this good now?

czgdp1807 commented:
There is a lot of similar code in the PR reducing which will be quite challenging. 
ping @Upabjojr Please provide your comments on this PR.

czgdp1807 commented:
ping @Upabjojr

czgdp1807 commented:
Will merge it by tomorrow if no objections raised.
